cupy/random/_bit_generator.pyx

```diff
@@ -0,0 +1,184 @@
+        # TODO(ecastill) port SeedSequence
+        self._seed_seq = numpy.random.SeedSequence(seed)
+        dev = cupy.cuda.Device()
+        self._current_device = dev.id
+
+    def random_raw(self, size=None, out=False):
+        raise NotImplementedError(
+            'Subclasses of `BitGenerator` must override `random_raw`')
+
+    def state_size(self):
+        """Maximum number of samples that can be generated at once
+        """
+        return 0
+
+    def _check_device(self):
+        if cupy.cuda.Device().id != self._current_device:
+            raise RuntimeError("This Generator state has been allocated in a different device")
+
+
+class _cuRANDGenerator(BitGenerator):
+    # Size is the number of threads that will be initialized
+    def __init__(self, seed=None, size=1024*100):
+        super().__init__(seed)
+        self._seed = self._seed_seq.generate_state(1, numpy.uint64)[0]
+        self._size = size
+        cdef ssize_t b_size = self._type_size() * size
+        self._state = cupy.zeros(b_size, dtype=numpy.int8)
+        ptr = self._state.data.ptr
+        cdef intptr_t state_ptr = <intptr_t>ptr
+        cdef uint64_t c_seed = <uint64_t>self._seed
+        cdef intptr_t _strm = stream.get_current_stream_ptr()
+        # Initialize the state
+        init_curand_generator(self.generator, state_ptr, self._seed, size, _strm)
+
+    def random_raw(self, size=None, out=False):
+        pass
+
+    def state(self):
+        self._check_device()
+        return self._state.data.ptr
+
+    def state_size(self):
+        return self._size
+
+    def _type_size(self):
+        return 0 
+
+class XORWOW(_cuRANDGenerator):
+    generator = CURAND_XOR_WOW  # Use The Enum
+
+    def _type_size(self):
+        return sizeof(curandState)
+
+class MRG32k3a(_cuRANDGenerator):
+    generator = CURAND_MRG32k3a
+
+    def _type_size(self):
+        return sizeof(curandStateMRG32k3a)
+
+
+class Philox4x3210(_cuRANDGenerator):
+    generator = CURAND_PHILOX_4x32_10
+
+    def _type_size(self):
+        return sizeof(curandStatePhilox4_32_10_t)
+
+
+class Generator:
+    def __init__(self, bit_generator):
+        self._bit_generator = bit_generator 
+
+    def integers(self, low, high, size, dtype=numpy.int32, endpoint=False):
+        cdef ndarray y
+
+        diff = high-low
+        if not endpoint:
+           diff -= 1
+
+        cdef uint64_t mask = (1 << diff.bit_length()) - 1
+        # TODO adjust dtype
+        if diff <= _UINT32_MAX:
+            dtype = numpy.uint32
+        elif diff <= _UINT64_MAX:
+            dtype = numpy.uint64
+        else:
+            raise ValueError(
+                'high - low must be within uint64 range (actual: {})'.format(diff))
+
+        y = ndarray(size if size is not None else (), dtype)
+
+        if dtype is numpy.uint32:
+           self._launch_distribution_kernel(interval_32, y, diff, mask)
+        else:
+           self._launch_distribution_kernel(interval_64, y, diff, mask)
```

Comment:
Nitpick: Seems like `interval_32`/`interval_64` are passed around as Python objects, despite that they're in fact C functions. If this is valid, perhaps we can just merge this part with the previous dtype check:
```python
if diff <= _UINT32_MAX:
    dtype = numpy.uint32
    func = interval_32
elif diff <= _UINT64_MAX:
    dtype = numpy.uint64
    func = interval_64
else:
    raise ...

y = ndarray(...)
self._launch_distribution_kernel(func, y, diff, mask)
```

Reply:
Looks like Travis has a hard time understanding `func = void (*)(...)` 😂